### PR TITLE
Remove tablet query plan field caching

### LINKED ISSUE #10210: Explore Reducing Scope of Tablet Query Plan Cache
### Feature Description

The tabletserver query plan cache will keep query plans around associated by their SQL statement string.

One of the functions of the query plan cache is that it can optionally cache field information. It appears that this is largely to improve the parsing performance of cached query plans because it allows parsing this information in the result packet. However, this is not safe in the face of schema changes, which can potentially catastrophically corrupt the results that are returned to the client in real world cases. Further, it doesn't appear to provide a major performance gain. Inspecting the code, the additional work is UTF-8 validation.

I put together a quick microbenchmark within the framework provided by `go/mysql/query_benchmark.go` of parsing a single-row result with 10 columns, each with 1-byte values, and the difference was around 6% (~750ns per result) for largely just parsing:

```cpu: Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz                                                                                                     
BenchmarkWantfieldsTrueOneRow-4           943192             12566 ns/op           1.59 MB/s
BenchmarkWantfieldsFalseOneRow-4          884964             11798 ns/op           1.70 MB/s                
```

The field cache itself adds meaningful complexity to the planning path and seems to already have some issues, such as where it holds a read lock during blocking I/O: https://github.com/vitessio/vitess/blob/8de235531d8af08dbe4f96ff6e63b24ad9de9481/go/vt/vttablet/tabletserver/query_engine.go#L328

There's probably reasonable justification for deprecating and removing the field cache given the correctness concerns and it's very marginal performance impact, which could pave the way to simplify the plan cache. So let's talk about the plan cache.

Calling the tabletservers planner a "planner" is really a bit of a misnomer, particularly compared to what's going on in vtgate. It's more of a "decorator."  To do its job, however, it does need to parse the SQL statement. I'm not entirely sure, but it does seem like that is likely the bulk of the work during the planning phase within the tabletserver. Parsing SQL is a pure function, so the same input SQL string will always produce the same result. The ideal caching scenario. I'd propose running some benchmarks where the planner is paired back to only caching the result of the SQL parsing and see if there's a meaningful difference. If not, it's probably worth saving the complexity and chance for introducing severe bugs on a very critical path.

### Use Case(s)

This covers all query execution within the tabletserver.

### PR DESCRIPTION
**Pull request by arvind-murty**
## Description

This PR removes tablet query plan field caching. The flag `enable_query_plan_field_caching` is now deprecated and can be removed in a future release.
Note: Prior to this PR, query consolidation was gated by the query plan field caching flag. Now, it will be active unless `--enable-consolidator` is set to false.

## Related Issue(s)

Fixes #10210 

## Checklist

-   [ ] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [ ] Documentation was added or is not required

## Deployment Notes

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1202058585380140

### CODE REVIEW

File: go/vt/vttablet/tabletserver/planbuilder/plan.go [hunk 1]
```diff
@@ -162,9 +162,6 @@ type Plan struct {
 	// Permissions stores the permissions for the tables accessed in the query.
 	Permissions []Permission
 
-	// FieldQuery is used to fetch field info
-	FieldQuery *sqlparser.ParsedQuery
-
```
Comment (arvind-murty) [inline go/vt/vttablet/tabletserver/planbuilder/plan.go:167]: When I tried to `make codegen`, I got the following errors:
```
% make codegen
2022/06/08 18:56:21 saved '/Users/arvind/go/src/vitess.io/vitess/go/vt/sqlparser/ast_clone.go'
2022/06/08 18:56:21 saved '/Users/arvind/go/src/vitess.io/vitess/go/vt/sqlparser/ast_visit.go'
2022/06/08 18:56:21 saved '/Users/arvind/go/src/vitess.io/vitess/go/vt/sqlparser/ast_rewrite.go'
2022/06/08 18:56:21 saved '/Users/arvind/go/src/vitess.io/vitess/go/vt/sqlparser/ast_equals.go'
2022/06/08 18:56:31 /Users/arvind/go/src/vitess.io/vitess/go/vt/vttablet/tabletserver/planbuilder/cached_size.go:63:17: cached.FieldQuery undefined (type *Plan has no field or method FieldQuery)
2022/06/08 18:56:31 /Users/arvind/go/src/vitess.io/vitess/go/vt/vttablet/tabletserver/cached_size.go:36:50: cached.Fields undefined (type *TabletPlan has no field or method Fields)
2022/06/08 18:56:31 /Users/arvind/go/src/vitess.io/vitess/go/vt/vttablet/tabletserver/cached_size.go:37:31: cached.Fields undefined (type *TabletPlan has no field or method Fields)
2022/06/08 18:56:31 failed to load packages
exit status 1
make: *** [sizegen] Error 1
```

I had to manually edit `cached_size.go` and remove the deleted `FieldQuery` and then run `make codegen` again.
This seems like a bug in `make sizegen`.

File: go/vt/vttablet/tabletserver/cached_size.go [hunk 2]
```diff
@@ -25,19 +25,12 @@ func (cached *TabletPlan) CachedSize(alloc bool) int64 {
...
 	size += cached.Plan.CachedSize(true)
 	// field Original string
 	size += hack.RuntimeAllocSize(int64(len(cached.Original)))
-	// field Fields []*vitess.io/vitess/go/vt/proto/query.Field
-	{
-		size += hack.RuntimeAllocSize(int64(cap(cached.Fields)) * int64(8))
-		for _, elem := range cached.Fields {
-			size += elem.CachedSize(true)
-		}
-	}
```
Comment (arvind-murty) [inline go/vt/vttablet/tabletserver/cached_size.go:40]: `make codegen` gave an error for this file as well.

File: go/vt/vttablet/tabletserver/query_engine_test.go [hunk 3]
```diff
@@ -192,14 +192,14 @@ func TestQueryPlanCache(t *testing.T) {
...
-	firstPlan, err := qe.GetPlan(ctx, logStats, firstQuery, false, 0, nil)
+	firstPlan, err := qe.GetPlan(logStats, firstQuery, false, 0, nil)
 	if err != nil {
 		t.Fatal(err)
 	}
 	if firstPlan == nil {
 		t.Fatalf("plan should not be nil")
 	}
-	secondPlan, err := qe.GetPlan(ctx, logStats, secondQuery, false, 0, nil)
+	secondPlan, err := qe.GetPlan(logStats, secondQuery, false, 0, nil)
```
Comment (arvind-murty) [inline go/vt/vttablet/tabletserver/query_engine_test.go:202]: This test is failing for me with the following error.
```
--- FAIL: TestQueryPlanCache (0.00s)
    query_engine_test.go:169: expected query plan cache to contain 1 entries, found 2
```
`firstQuery` and `secondQuery` are not the same, so I would expect the plan cache to contain both. Why are we expecting only one?

File: go/vt/vttablet/tabletserver/planbuilder/plan.go (hunk 1 above)
Comment (deepthi) [reply go/vt/vttablet/tabletserver/planbuilder/plan.go:167]: cc @vmg

Comment (harshit-gangal) [reply go/vt/vttablet/tabletserver/planbuilder/plan.go:167]: This is the limitation of generated files in go (or at least in how we do generated files).
This is not a bug for now.

May be @vmg can add details better.

**Comment by frouioui** [comment]
Regarding the performance improvement, there is no notable improvement in arewefastyet. The comparison is available [here](https://benchmark.vitess.io/compare?r=9cbbf6faa46acdcd780ba46fb48a6471103f4a14&c=ac8809d29bcfef3037c1f78f36d0dcba1fdee5db).

However, I would be curious to see how [this benchmark](https://github.com/vitessio/vitess/compare/main...rbranson:benchmark-wantfields) performs before and after the changes.

File: go/vt/vttablet/tabletserver/query_engine.go [hunk 4]
```diff
@@ -289,10 +281,7 @@ func (qe *QueryEngine) Close() {
 }
 
 // GetPlan returns the TabletPlan that for the query. Plans are cached in a cache.LRUCache.
-func (qe *QueryEngine) GetPlan(ctx context.Context, logStats *tabletenv.LogStats, sql string, skipQueryPlanCache bool, reservedConnID int64, te *TxEngine) (*TabletPlan, error) {
-	span, ctx := trace.NewSpan(ctx, "QueryEngine.GetPlan")
-	defer span.Finish()
```
[the remark this answers is not in the document]
Comment (deepthi) [reply go/vt/vttablet/tabletserver/query_engine.go:294]: With the removal of the RPC there is no need for a context any more. `NewSpan` expects a context.
What is the other option?

**Comment by arvind-murty** [comment]
> Regarding the performance improvement, there is no notable improvement in arewefastyet. The comparison is available [here](https://benchmark.vitess.io/compare?r=9cbbf6faa46acdcd780ba46fb48a6471103f4a14&c=ac8809d29bcfef3037c1f78f36d0dcba1fdee5db).
> 
> However, I would be curious to see how [this benchmark](https://github.com/vitessio/vitess/compare/main...rbranson:benchmark-wantfields) performs before and after the changes.

`wantfields` benchmark on main:
```
BenchmarkWantfieldsTrueOneRow-4    	   52132	     23221 ns/op	   0.86 MB/s
BenchmarkWantfieldsFalseOneRow-4   	   53290	     23118 ns/op	   0.87 MB/s
```

`wantfields` benchmark on this branch:
```
BenchmarkWantfieldsTrueOneRow-4    	   48429	     23015 ns/op	   0.87 MB/s
BenchmarkWantfieldsFalseOneRow-4   	   49940	     25488 ns/op	   0.78 MB/s
```

**Comment by frouioui** [comment]
@arvind-murty thank you for the benchmarking results. Just to confirm how significant these results are do you mind running `benchstat` (https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) to get the p-value.

**Comment by systay** [comment]
Here you go:

```
name                     old time/op   new time/op   delta
WantfieldsTrueOneRow-4    23.2µs ± 0%   23.0µs ± 0%   ~     (p=1.000 n=1+1)
WantfieldsFalseOneRow-4   23.1µs ± 0%   25.5µs ± 0%   ~     (p=1.000 n=1+1)

name                     old speed     new speed     delta
WantfieldsTrueOneRow-4   860kB/s ± 0%  870kB/s ± 0%   ~     (p=1.000 n=1+1)
WantfieldsFalseOneRow-4  870kB/s ± 0%  780kB/s ± 0%   ~     (p=1.000 n=1+1)
```

**Comment by arvind-murty** [comment]
> > @arvind-murty thank you for the benchmarking results. Just to confirm how significant these results are do you mind running `benchstat` (https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) to get the p-value.
> 
> EDIT: Let's run the benchmarks with the `-count` argument greater than `1`, i.e. `-count=10`. For both branches (`main` and `am-wantfields`). After running the benchmarks on a branch, you will want to save the output to a file, so `benchstat` can use the output again later, like so:
> 
> ```
> benchstat results_main.txt results_new_branch.txt
> ```

Here are the results with `-count=10`:

```
% benchstat main.txt branch.txt
name                     old time/op    new time/op    delta
ParallelShortQueries-4     14.7µs ±45%    14.0µs ±52%     ~     (p=0.579 n=10+10)
ParallelMediumQueries-4    66.3µs ±76%    34.9µs ± 2%  -47.34%  (p=0.000 n=9+10)
ParallelRandomQueries-4    79.4µs ±28%    63.0µs ±24%  -20.72%  (p=0.004 n=10+10)
WantfieldsTrueOneRow-4     31.9µs ±39%    24.3µs ± 3%  -23.86%  (p=0.022 n=10+9)
WantfieldsFalseOneRow-4    24.8µs ± 2%    23.8µs ± 4%   -4.17%  (p=0.000 n=9+10)

name                     old speed      new speed      delta
ParallelShortQueries-4   1.50MB/s ±35%  1.58MB/s ±37%     ~     (p=0.540 n=10+10)
ParallelMediumQueries-4   251MB/s ±50%   470MB/s ± 2%  +87.35%  (p=0.000 n=10+10)
WantfieldsTrueOneRow-4    656kB/s ±31%   822kB/s ± 3%  +25.34%  (p=0.028 n=10+9)
WantfieldsFalseOneRow-4   807kB/s ± 2%   841kB/s ± 4%   +4.26%  (p=0.000 n=9+10)

```

**Comment by systay** [comment]
@arvind-murty 👋 you need to merge main into your branch. There are a couple of conflicts that need to be resolved

**Comment by arvind-murty** [comment]
One of the test failures is:
```
=== RUN   TestAggregateTypes
[321](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:322)
    aggregation_test.go:64: Query: select ascii(val1) as a, count(*) from aggr_test group by a (-want +got):
[322](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:323)
          strings.Join({
[323](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:324)
          	"[[INT",
[324](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:325)
        - 	"64",
[325](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:326)
        + 	"32",
[326](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:327)
          	"(65) INT64(1)] [INT",
[327](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:328)
        - 	"64",
[328](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:329)
        + 	"32",
[329](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:330)
          	"(69) INT64(1)] [INT",
[330](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:331)
        - 	"64",
[331](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:332)
        + 	"32",
[332](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:333)
          	"(97) INT64(1)] [INT",
[333](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:334)
        - 	"64",
[334](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:335)
        + 	"32",
[335](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:336)
          	"(98) INT64(1)] [INT",
[336](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:337)
        - 	"64",
[337](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:338)
        + 	"32",
[338](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:339)
          	"(99) INT64(2)] [INT",
[339](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:340)
        - 	"64",
[340](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:341)
        + 	"32",
[341](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:342)
          	"(100) INT64(1)] [INT",
[342](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:343)
        - 	"64",
[343](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:344)
        + 	"32",
[344](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:345)
          	"(101) INT64(1)]]",
[345](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:346)
          }, "")
[346](https://github.com/vitessio/vitess/runs/7066655920?check_suite_focus=true#step:10:347)
```

@frouioui I see that you have a TODO in this test that seems to correspond to this diff. Any idea why this is failing?
https://github.com/vitessio/vitess/blob/main/go/test/endtoend/vtgate/queries/aggregation/aggregation_test.go#L61-L66

**Comment by arvind-murty** [comment]
> @frouioui I see that you have a TODO in this test that seems to correspond to this diff. Any idea why this is failing? https://github.com/vitessio/vitess/blob/main/go/test/endtoend/vtgate/queries/aggregation/aggregation_test.go#L61-L66

Looks like the changes in this PR actually fixed your TODO, so I have changed the test to run with `mcmp`.

**Comment by deepthi** [comment]
The last change fixes an interesting bug resulting from field caching. I added some logging, and looks like the field info returned from MySQL is different when you have a `where 1 !=1` resulting in an empty result and when you have actual results.
Check this out, here `Plan Fields` are from executing the field query and `Result Fields` are from executing the actual query with `wantfields` set to true. 
```
E0629 22:24:48.645320   25877 query_engine.go:351] Field Query: select ascii(val1) as a, count(*), weight_string(ascii(val1)) from aggr_test where 1 != 1 group by a, weight_string(ascii(val1))
E0629 22:24:48.645335   25877 query_engine.go:352] Plan fields: [name:"a" type:INT64 column_length:3 charset:63 flags:32896 name:"count(*)" type:INT64 column_length:21 charset:63 flags:32897 name:"weight_string(ascii(val1))" type:VARBINARY column_length:8 charset:63 decimals:31 flags:128]
E0629 22:24:48.645683   25877 query_executor.go:633] Result fields: [name:"a" type:INT32 org_name:"a" column_length:3 charset:63 flags:32768 name:"count(*)" type:INT64 column_length:21 charset:63 flags:32769 name:"weight_string(ascii(val1))" type:VARBINARY org_name:"weight_string(ascii(val1))" column_length:8 charset:63 flags:32896]
```

File: go/vt/vttablet/tabletserver/query_engine_test.go [hunk 5]
```diff
@@ -188,18 +188,20 @@ func TestQueryPlanCache(t *testing.T) {
 	ctx := context.Background()
 	logStats := tabletenv.NewLogStats(ctx, "GetPlanStats")
 	if cache.DefaultConfig.LFU {
-		qe.SetQueryPlanCacheCap(1024)
+		// this cache capacity is in bytes
```
Comment (systay) [inline go/vt/vttablet/tabletserver/query_engine_test.go:191]: nice commenting. it's a good habit, adding comments for things that made you scratch your head :)

File: go/vt/vttablet/tabletserver/query_executor.go [hunk 6]
```diff
@@ -623,27 +614,44 @@ func (qre *QueryExecutor) execNextval() (*sqltypes.Result, error) {
...
-	if qre.tsv.qe.enableQueryPlanFieldCaching && qre.plan.Fields != nil {
-		result, err := qre.qFetch(qre.logStats, qre.plan.FullQuery, qre.bindVars)
-		if err != nil {
-			return nil, err
+	sql, sqlWithoutComments, err := qre.generateFinalSQL(qre.plan.FullQuery, qre.bindVars)
+	if err != nil {
+		return nil, err
+	}
+	// Check tablet type.
+	if qre.shouldConsolidate() {
```
Comment (systay) [inline go/vt/vttablet/tabletserver/query_executor.go:622]: if I understand the changes correctly, this is new code, right? Do you think we should add a unit test for this part of the code, or are we already covered?

Comment (systay) [reply go/vt/vttablet/tabletserver/query_executor.go:622]: ah, I see that you have just inlined `qFetch`. Ignore my comment☝️

File: go/vt/vttablet/tabletserver/query_engine.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (harshit-gangal) [reply go/vt/vttablet/tabletserver/query_engine.go:294]: I am not able to find the RPC that is removed. Also, GetPlan is called from almost all the different Execute apis.
I think we should keep the span and the context here.

**Review by harshit-gangal — COMMENTED** [review]
Overall looks good, only one comment that needs to be addressed.

File: go/vt/vttablet/endtoend/queries_test.go [hunk 7]
```diff
@@ -1753,15 +1720,15 @@ var TestQueryCases = []framework.Testable{
 				Name:  "specific column",
 				Query: "select eid from vitess_a where 1 != 1",
 				Rewritten: []string{
-					"select eid from vitess_a where 1 != 1",
+					"select eid from vitess_a where 1 != 1 limit 10001",
```
[the remark this answers is not in the document]
Comment (harshit-gangal) [reply go/vt/vttablet/endtoend/queries_test.go:1723]: followed the code ignore this.

**Review by harshit-gangal — APPROVED** [review]

**Comment by deepthi** [comment]
Looks like all feedback has been addressed. DCO has been acting up, so I've manually verified that the last commit has been correctly signed-off.

**Comment by deepthi** [comment]
And right after I merged this, I realized it probably needs a website docs update.
Reviewers, please be mindful of the checklist.

**Pull request merged**